### ISSUE TITLE
Full FP16 Dreambooth for SDXL does not work

### ISSUE THREAD

**Issue by DarkViewAI** [bmaltais/kohya_ss#2652]
Either produces nans, or doesn't train at all, all the sample images are the same.

**Comment by b-fission** [comment]
Did you enable "No half VAE" in Dreambooth? If the option is not showing, you'll have to specify `--no_half_vae` as an additional parameter.

**Comment by DarkViewAI** [comment]
@b-fission yes, everything works fine with bf16 but its not with full-fp16

**Comment by WarAnakin** [comment]
i know the setting is called full fp16 pr bf, i'd just like to point this is a misrepresentation of what that parameter does. You are already training in FP or BF, that setting has nothing to do with full precision, it just includes some gradients.
There is no such thing as FULL FP16, it's like saying full half-precision. Full precision = 32, half precision = 16

![image](https://github.com/user-attachments/assets/1dd45891-13c0-4626-9b5c-c01bf6159d84)

**Comment by b-fission** [comment]
> that setting has nothing to do with full precision, it just includes some gradients. There is no such thing as FULL FP16, it's like saying full half-precision. Full precision = 32, half precision = 16

I'm pretty sure it's referring to "full training in half-precision" as opposed to the default of training in mixed-precision.

**Comment by WarAnakin** [comment]
I believed that for a long time as well, but there is no such thing as full training in half precision. The "mixed precision" parameter stands for the compute type to be used during training: FP32, FP16 and BF16. The other option is for saving the weights as FP32, FP16, BF16.
![image](https://github.com/user-attachments/assets/204c5887-1377-4dc7-9f9d-64a422361b6d)
![image](https://github.com/user-attachments/assets/d389a056-c788-44c3-986f-ff847cf427b5)
The FULL BF, FULL FP setting has nothing to do with changing the floating point compute operation. If you enable it, it's actually using less vram than disabled. Since when does higher precision computing use less resources ? I hope you see how some things just don't add up and this is because people take what they see at face value. The gui is misrepresenting what is actually happening in the back-end, this is why so many people have a hard time understanding and getting good results.

**Comment by b-fission** [comment]
> The FULL BF, FULL FP setting has nothing to do with changing the floating point compute operation. If you enable it, it's actually using less vram than disabled. Since when does higher precision computing use less resources ?

This is where I think the misunderstanding is coming from. Enabling the Full fp16/bf16 option does not mean it's training in full-precision/fp32.

I'll use the sdxl training script for reference.
https://github.com/kohya-ss/sd-scripts/blob/25f961bc779bc79aef440813e3e8e92244ac5739/sdxl_train.py#L388-L403

Some elements like the unet get loaded in fp32 _by default_ even when mixed-precision is set to fp16/bf16 or Off. You can verify the datatype by pasting this `print` statement right after the part that I linked. 

```python
print(f"unet is {unet.dtype} ... te1 is {text_encoder1.dtype}")
```

If you observe the output of that line, you should see that `unet` only becomes half-precision when Full fp16/bf16 is enabled, which explicitly converts it from 32-bit->16-bit, leading to lower vram usage.

**Comment by WarAnakin** [comment]
thx for high-lighting that, it was an eye opener.
